receiver.go

```diff
@@ -53,30 +54,46 @@ func newReceiver(context *Context, conn *websocket.Conn) *receiver {
 	}
 }
 
-func (r *receiver) receiveLoop() error {
+func (r *receiver) getParseError() error {
+	r.parseErrorLock.Lock()
+	defer r.parseErrorLock.Unlock()
+	return r.parseError
+}
+
+func (r *receiver) setParseError(err error) {
+	r.parseErrorLock.Lock()
+	defer r.parseErrorLock.Unlock()
+	r.parseError = err
+}
+
+func (r *receiver) receiveLoop() (err error) {
 	defer func() {
 		atomic.AddInt32(&r.activeGoroutines, -1)
 	}()
 	go r.parseLoop()
 
 	atomic.AddInt32(&r.activeGoroutines, 1)
 	for {
+		if err = r.getParseError(); err != nil {
```

Comment:
Acquiring a mutex lock every iteration of the receiveLoop seems like a lot of overhead to check for what's only a one-time error.  I thought there would be better synchronization patterns for one-time communication between loops (i.e. when the parse loop raises error, it's terminating) - maybe just closing a channel?  Or did you find that the overhead of something like a select/default is higher than acquiring and releasing the mutex?

Reply:
I originally had a channel with the error (of size 1 to avoid deadlocking) but assumed a lower level mutex was less overhead than a channel.

I'll benchmark and determine that for sure.

Reply:
Yep, I was wrong. Reading an error over a buffered channel is twice as fast as a mutex. I'll swap it back.

```
BenchmarkChannelMutex/chan_error-12         	223629256	         5.17 ns/op
BenchmarkChannelMutex/error_with_mutex-12   	100000000	        10.7 ns/op
BenchmarkChannelMutex/atomicValue_error-12  	150909745	         7.90 ns/op
```